### ISSUE TITLE
Bug: CMB2 should be run as late as possible

### ISSUE THREAD

**Issue by superdummy** [WebDevStudios/CMB2#338]
Problem: 
CMB2 requires to add/define the metaboxes (in the backend) as follows:
`add_filter( 'cmb2_meta_boxes', array ($this, 'metaboxes' ) );`

My Metaboxes are very heavy, in that they look up a lot of data what to display, when to display and so on (not really be accomplishable by using the cmb2 callbacks in the fields)

So I need to limit the definition of the Metaboxes to only a specific screen. Otherwise my logic is triggered on all (!) pages in the backend. As the first statments shows, when using this hook, the metaboxes will be added generally for every request in the backend. But I need to limit it to a specific screen. 

Therefore I use:
`add_action( 'current_screen', array($this, 'registerAdminPageHooks') );`

This hook is the first hook when the "current_screen" object ist populated and allows to check on which page you are in the backend. But when I use this hook (and test for the specific Screen/CPT) to add the metaboxes (like in the first statements). Then the Metaboxes are not shown. It seems, that this hook is to late for CMB2. But this hook is extremely important to be able to check on which page you are and then, basedn on this logik either define the metaboxes or not define the metaboxes. Globally defining Metaboxes for the whole backend is a huge waste of resources...

I think this is a very common problem and use case:  And CMB2 should be fixed to run "as late as possible" so that I can use current_screen hook, to actually define the metaboxes there.

Thanks

**Comment by austinjreilly** [comment]
I'm not sure when this changed, but I think you're supposed to use `cmb2_init` as your hook, not `cmb2_meta_boxes` (see https://github.com/WebDevStudios/CMB2/wiki/Basic-Usage#create-a-metabox). Does that address your issue at all?

**Comment by jtsternberg** [comment]
First, the `'cmb2_meta_boxes'` is deprecated, and you should be [using the new metabox/field registration API](https://github.com/WebDevStudios/CMB2/wiki/Basic-Usage#create-a-metabox). Second, to do your logic, you should use the ['show_on_cb' callback parameter on the metabox registration array](https://github.com/WebDevStudios/CMB2/blob/master/example-functions.php#L84) to do any kind of logic for displaying your metabox. This off-loads the logic to the last opportunity it can.

**Comment by superdummy** [comment]
Thanks for your advice and help!

The Filter/Action seems to be have changed in the last 9month. The filter  `cmb2_meta_boxes` I used, was base on a 6-9 old version of the example provided at that time.

Yes it seems, what I wanted to do can be acomplished with the show_on_cb callback. 

After changing the filter/action to  `cmb2_init` I did a retest, and it did not work either. Just as a thought: I personally would think it adds a lot of "bonus points" when the cmb2_init would be triggered extremely late, the latest as possible (so one must not necessarily depend on callbacks which add in my opinion some overhead and are not that intuitive as doing this directly in the code).But yes, using the callback is an alternative and works. 

Thanks for your help and adivce.
